File: 1760-minimum-limit-of-balls-in-a-bag/1760-minimum-limit-of-balls-in-a-bag.cpp
class Solution {
private:
    // Helper function to check if a distribution is possible for a given maxBallsInBag
    bool isPossible(vector<int>& nums, int maxBallsInBag, int maxOperations){
        int totalOperations = 0;
        
        // Iterate through each bag in the array
        for (int num : nums){
            // Calculate the number of operations needed to split this bag
            totalOperations += ceil((double)num / maxBallsInBag) - 1;
            
            // Total operations exceed maxOperations, return false
            if (totalOperations > maxOperations)
                return false;
        }
        
        // We can split the balls within the allowed operations, return true
        return true;
    }
    
public:
    int minimumSize(vector<int>& nums, int maxOperations) {
        // Binary search bounds
        int left = 1, right = *max_element(nums.begin(), nums.end());
        
        // Perform binary search to find the optimal maxBallsInBag
        while (left < right){
            int mid = left + ((right - left) >> 1);
            
            // Check if a valid distribution is possible within the current middle value
            if (isPossible(nums, mid, maxOperations))
                right = mid;    // If possible, try a smaller value (shift right to middle)
            else
                left = mid + 1; // If not possible, try a larger value (shift left to middle + 1)
        }
        
        // Return the smallest possible value for maxBallsInBag
        return left;
    }
};

File: 1760-minimum-limit-of-balls-in-a-bag/1760-minimum-limit-of-balls-in-a-bag.py
class Solution:
    def minimumSize(self, nums: List[int], maxOperations: int) -> int:
        # Helper function to check if a distribution is possible for a given max_balls_in_bag
        def is_possible(max_balls_in_bag: int) -> bool:
            total_ops = 0
            
            # Iterate through each bag in the array
            for num in nums:
                # Calculate the number of operations needed to split this bag
                total_ops += ceil(num / max_balls_in_bag) - 1
                
                # If total operations exceed max_opertaions, return False
                if total_ops > maxOperations:
                    return False
            
            # We can split the balls within the allowed operations, return True
            return True 
        
        # Binary search bounds
        left, right = 1, max(nums)
        
        # Perform binary search to find the optimal max_balls_in_bag
        while left < right:
            mid = (left + right) >> 1
            
            # Check if a valid distribution is possible with the current middle value
            if is_possible(mid):
                # If possible, try a smaller value (shift right to middle)
                right = mid
            else:
                # If not possible, tru a larger value (shift left to middle + 1)
                left = mid + 1
                
        # Return the smallest possible value for max_balls_in_bag
        return left